ReactiveCocoa/Swift/Property.swift

```diff
@@ -5,69 +5,182 @@ import enum Result.NoError
 public protocol PropertyType {
 	associatedtype Value
 
-	/// The current value of the property.
-	var value: Value { get }
-
 	/// A producer for Signals that will send the property's current value,
-	/// followed by all changes over time.
+	/// followed by all changes over time, then complete when the property has
+	/// deinitialized.
 	var producer: SignalProducer<Value, NoError> { get }
 
 	/// A signal that will send the property's changes over time.
 	var signal: Signal<Value, NoError> { get }
+
+	/// Performs an arbitrary action using the current value of the
+	/// variable. Conforming types may optionally guarantee atomicity.
+	///
+	/// Returns the result of the action.
+	func withValue<Result>(@noescape action: Value throws -> Result) rethrows -> Result
 }
 
-/// A read-only property that allows observation of its changes.
-public struct AnyProperty<Value>: PropertyType {
+/// Represents an observable property that can be mutated directly.
+///
+/// Only classes can conform to this protocol, because instances must support
+/// weak references (and value types currently do not).
+public protocol MutablePropertyType: class, PropertyType {
+	func modify(@noescape action: Value throws -> Value) rethrows -> Value
+}
+
+extension PropertyType {
+	public var value: Value {
+		return withValue { $0 }
+	}
+
+	/// Maps the current value and all subsequent values as a new property.
+	///
+	/// A mapped property would retain its source property.
+	@warn_unused_result(message="Did you forget to use the composed property?")
+	public func map<U>(transform: Value -> U) -> AnyProperty<U> {
+		return withValue { currentValue in
+			let producer = SignalProducer(signal: signal)
+				.map(transform)
+				.on(completed: { self })
+
+			return AnyProperty(initialValue: transform(currentValue), producer: producer)
+		}
+	}
+
+	/// Combines the current value and the subsequent values of two `Property`s in
+	/// the manner described by `Signal.combineLatestWith:`, and applys the supplied
+	/// transform on the combined pair.
+	///
+	/// A combined property would retain its source properties.
+	@warn_unused_result(message="Did you forget to use the composed property?")
+	public func combineLatest<P: PropertyType, U>(with other: P, combining transform: (Value, P.Value) -> U) -> AnyProperty<U> {
+		return withValue { currentValue in
+			return other.withValue { currentOtherValue in
```

Comment:
Hmm, I now see that this is risky. If one composes `propertyA` with `propertyB`, and composes another in the reverse order at the same time, it has a chance to deadlock.